src/development/verkle-replay.md

```diff
@@ -0,0 +1,140 @@
+
+## Witnesses
+
+This section explores multiple dimensions of the execution witness design.
+
+### Size
+
+Research has been conducted, in order to determine the best structure for a witness. We tried four methods:
+
+ * The first one, as specified in the [current version of the consensus spec](https://github.com/ethereum/consensus-specs/blob/dev/specs/_features/eip6800/beacon-chain.md#executionwitness), referred to as **type 0**. This approach uses the SSZ type `Optional[T]` , that is yet to be officially included in the spec.
+ * A first variation of the spec, in which prestate values are grouped together in their own list, and poststate values are grouped in their own list as well. This is referred to as **type 1**.
+
+```python
+class SuffixStateDiffs(Container):
+    suffixes: bytes
+    # None means not currently present
+    current_values: List[bytes]
+    # None means value not updated
+    new_values:     List[bytes]
+```
+Type 1 serialization container
+
+ * A second variation of the spec, in which updates, insertions and reads are grouped in their own separate lists. The suffixes for each of these lists are also grouped as their own byte lists. This is referred to as **type 2**.
+
+```python
+class SuffixStateDiffs(Container):
+    updated_suffixes: bytes
+    updated_currents: List[Bytes32]
+    updated_news: List[Bytes32]
+
+    inserted_suffixes: bytes
+    inserted_news: List[Bytes32]
+
+    read_suffixes: bytes
+    read_currents: List[Bytes32]
+
+    missing_suffixes: bytes
+```
+Type 2 serialization container
+
+ * A final variation, where update, reads, inserts and proof-of-absences are grouped, along with their suffixes, into their own lists. This is referred to as **type 3**.
+
+```python
+class UpdateDiff(Container):
+    suffix: byte
+    current: Bytes32
+    new: Bytes32
+
+class ReadDiff(Container):
+    suffix: byte
+    current: Bytes32
+
+class InsertDiff(Container):
+    suffix: byte
+    new: Bytes32
+
+class MissingDiff(Container):
+    suffix: byte
+class StemStateDiff(Container):
+    stem: Bytes31
+
+    updates: List[UpdateDiff]
+    reads: List[ReadStateDiff]
+    insert: List[InsertDiff]
+    missing: List[MissingStateDiff]
+```
+Type 3 serialization container
+
+Here are our findings after replaying data:
+
+![image.png](./assets/total.png)
+
+One can see that the spec is much less efficient than the other types. This is due to the inefficiency of storing `Optional[T]`. As a result, we didn’t include this method in the rest of the analysis.
+
+When only considering witnesses that were produced after the whole state was [converted to verkle](../state-conversion/intro.md), this is what is found:
+
+![image.png](./assets/compare_post_transition.png)
+
+The maximum and minimum observed sizes for each type, are summarized in the following table:
```

Comment:
Still not sure about the "minimum observed":
![image](https://github.com/user-attachments/assets/e809bc61-8801-4404-b63c-a6105eec12ff)

there are measurements where the minimum is almost zero. That was my point in my original review comment. 

BTW, I think saying the "minimum" is somewhat irrelevant exactly for the point I'm making. It's always zero or close to zero for mostly empty blocks.